File: 백준/Gold/2668. 숫자고르기/숫자고르기.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];
rl.on("line", (line) => {
  input.push(parseInt(line));
}).on("close", () => {
  const n = input[0];
  const answer = new Set();
  for (let index = 1; index < n + 1; index++) {
    start = index;
    // 체크해야한다
    if (!answer.has(start)) {
      const check = [];
      check[start] = true;
      now = input[start];
      const tmp = [start];
      while (true) {
        // 시작점과 같다면 가능함
        if (now == start) {
          // 정답에 추가
          answer.add(...tmp);
          break;
        } else {
          // 시작점이 아닌 곳으로 도착
          if (!!check[now]) break;
          tmp.push(now);
          check[now] = true;
          now = input[now];
        }
      }
    }
  }
  const ans = [...answer];
  console.log(ans.length);
  ans.sort(function (a, b) {
    return a - b;
  });
  for (const an of ans) {
    console.log(an);
  }
  process.exit();
});

File: 백준/Gold/2668. 숫자고르기/숫자고르기.java
import java.io.*;
import java.util.*;

public class Main {
    static StringTokenizer st;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int[] arr = new int[n+1];
        for (int i = 1; i < n+1; i++) {
            st = new StringTokenizer(br.readLine());
            arr[i] = Integer.parseInt(st.nextToken());
        }
        HashSet<Integer> answerSet = new HashSet<>();
        for (int i = 1; i < n+1; i++) {
            if (answerSet.contains(i)) continue;
            HashSet<Integer> tmp = new HashSet<>();
            int start = i;
            int now = i;
            while (!tmp.contains(now)) {
                tmp.add(now);
                now = arr[now];
            }
            if (now == start) {
                answerSet.addAll(tmp);
            }
        }
        Integer[] answer = answerSet.toArray(new Integer[0]);
        Arrays.sort(answer);
        System.out.println(answer.length);
        for (int ans: answer) {
            System.out.println(ans);
        }
    }
}